examples/vision/supervised-contrastive-learning.py

```diff
@@ -0,0 +1,206 @@
+"""
+Title: Supervised Contrastive Learning
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2020/11/01
+Last modified: 2020/11/01
+Description: Using supervised contrastive learning for image classification.
+
+
+## Introduction
+
+[Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362)
+(Prannay Khosla et al.) is a training methodology that outperforms
+ cross-entropy on supervised learning tasks.
+
+Essentially, training an image classification model with Supervised Contrastive
+Learning is performed in two phases:
+
+1. Pre-training an encoder to generate feature vectors for input images such
+that feature vectors of images in the same class will be more similar compared
+feature vectors of images in other classes.
+2. Training a classifier on top of the frozen encoder.
+
+## Setup
+"""
+
+"""shell
+pip install -q tensorflow-addons
+"""
+
+!pip install -q tensorflow-addons
+
+import tensorflow as tf
+import tensorflow_addons as tfa
+import numpy as np
+import keras 
+from keras import layers
+from keras.preprocessing.image import ImageDataGenerator
+
+"""## Prepare the data"""
+
+num_classes = 10
+input_shape = (32, 32, 3)
+
+# Load the train and test data splits
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+
+# Display shapes of train and test datasets
+print(f'x_train shape: {x_train.shape} - y_train shape: {y_train.shape}')
+print(f'x_test shape: {x_test.shape} - y_test shape: {y_test.shape}')
+
+"""## Using image data augmentation"""
+
+data_generator = ImageDataGenerator(
```

Comment:
ImageDataGenerator is legacy, don't use it. Instead, use preprocessing layers:

- `Normalization`: https://keras.io/api/layers/preprocessing_layers/core_preprocessing_layers/normalization/
- data augmentation layers (see example https://keras.io/examples/vision/image_classification_from_scratch/#using-image-data-augmentation)

Reply:
@fchollet - I tried using these APIs, but the performance (train and test) wasn't as good as using the ImageDataGenerator.

Basically I did the following:

```
data_augmentation = keras.Sequential(
    [
      layers.experimental.preprocessing.Normalization(),
      layers.experimental.preprocessing.RandomFlip("horizontal"),
      layers.experimental.preprocessing.RandomRotation(0.2),
      layers.experimental.preprocessing.RandomWidth(0.2),
      layers.experimental.preprocessing.RandomHeight(0.2)
    ]
)

def create_encoder():
  resnet = keras.applications.ResNet50V2(
      include_top=False,
      weights=None,
      input_shape=input_shape,
      pooling='avg')

  inputs = keras.Input(shape=input_shape)
  augmented = data_augmentation(inputs)
  outputs = resnet(augmented)
  model = keras.Model(inputs=inputs, outputs=outputs, name="cifar10-encoder")
  return model

encoder = create_encoder()
encoder.summary()

...

history = classifier.fit(
  x=x_train, 
  y=y_train, 
  batch_size=batch_size)
```

The test accuracy doesn't reach 70% in the baseline model. You can validate this using this [Colab](https://colab.research.google.com/drive/1z3EzFqIhZIIfnkQMulEIZRNJVq92eCYk#scrollTo=Fnt7P8C05mzs).

Do you have any suggestions around this?

Reply:
There are several considerations:

- `Normalization` needs its state to be set. You can do this by calling `layer.adapt(x_train)` after creating the layer.
- The image data augmentation may not be the same. Make sure to visualize the output images and adjust the amount of random rotation / etc accordingly. The meaning of the arguments has changed between the 2 APIs.

Reply:
Fixed. Thanks for the guidance!